tools/dds/dds-sniffer/dds-sniffer.cpp

```diff
@@ -144,159 +159,410 @@ void Sniffer::run( uint32_t seconds )
+        {
+            print_participant_discovered( info, false );
+        }
+        _discovered_participants.erase( info.info.m_guid );
+        break;
+    }
+}
 
-            if( _subscriber == nullptr )
-            {
-                return;
-            }
+void dds_sniffer::on_type_information_received( eprosima::fastdds::dds::DomainParticipant *,
+                                                const eprosima::fastrtps::string_255 topic_name,
+                                                const eprosima::fastrtps::string_255 type_name,
+                                                const eprosima::fastrtps::types::TypeInformation & type_information )
+{
+    std::cout << "Topic '" << topic_name << "' with type '" << type_name << "' received" << std::endl;
+}
+
+void dds_sniffer::save_topic_writer( const eprosima::fastrtps::rtps::WriterDiscoveryInfo & info )
+{
+    std::string topic_name = info.info.topicName().c_str();
+    _discovered_topics[topic_name].writers.insert( info.info.guid() );
+    save_max_indentation( std::move( topic_name ) );
+}
+
+void dds_sniffer::remove_topic_writer( const eprosima::fastrtps::rtps::WriterDiscoveryInfo & info )
+{
+    auto topic_entry = _discovered_topics[info.info.topicName().c_str()];
+    auto writer_iter = topic_entry.writers.find( info.info.guid() );
+    if( writer_iter != topic_entry.writers.end() )
+    {
+        topic_entry.writers.erase( writer_iter );
+        if( topic_entry.writers.empty() && topic_entry.readers.empty() )
+        {
+            _discovered_topics.erase( info.info.topicName().c_str() );
         }
+    }
+}
+
+void dds_sniffer::save_topic_reader( const eprosima::fastrtps::rtps::ReaderDiscoveryInfo & info )
+{
+    std::string topic_name = info.info.topicName().c_str();
+    _discovered_topics[topic_name].readers.insert( info.info.guid() );
+    save_max_indentation( std::move( topic_name ) );
+}
 
-        eprosima::fastdds::dds::Topic * topic = _participant->create_topic( topic_name.c_str(),
-                                                                            m_type->getName(),
-                                                                            eprosima::fastdds::dds::TOPIC_QOS_DEFAULT );
-        if( topic == nullptr )
+void dds_sniffer::remove_topic_reader( const eprosima::fastrtps::rtps::ReaderDiscoveryInfo & info )
+{
+    auto topic_entry = _discovered_topics[info.info.topicName().c_str()];
+    auto reader_iter = topic_entry.readers.find( info.info.guid() );
+    if( reader_iter != topic_entry.readers.end() )
+    {
+        topic_entry.readers.erase( reader_iter );
+        if( topic_entry.writers.empty() && topic_entry.readers.empty() )
         {
-            return;
+            _discovered_topics.erase( info.info.topicName().c_str() );
         }
+    }
+}
 
-        eprosima::fastdds::dds::DataReader * reader = _subscriber->create_datareader( topic, _readerQoS, this );
+void dds_sniffer::save_max_indentation( const std::string&& str )
```

Comment:
Why the `&&`? A regular reference would do, and a `std::move` wouldn't be needed...